src/cmakeTaskProvider.ts

```diff
@@ -237,12 +250,12 @@ export class CMakeTaskProvider implements vscode.TaskProvider {
             } else {
                 // Search for the matching default task.
                 const defaultTask: CMakeTask[] = matchingTargetTasks.filter(task => task.isDefault);
-                if (defaultTask.length === 1) {
+                if (defaultTask.length >= 1) {
                     return defaultTask[0];
                 } else {
                     // Search for the matching existing task.
                     const existingTask: CMakeTask[] = matchingTargetTasks.filter(task => !task.isTemplate);
```

Comment:
I don't see a `task.isTemplate` in tasks after you fetch tasks using` vscode.tasks.fetchTasks`. You have to distinguish between the tasks that are fetched from the tasks.json file and the tasks that the task provider provides as a template.

Reply:
@elahehrashedi
Yes. I mentioned this problem already.
https://github.com/microsoft/vscode-cmake-tools/issues/2936#issuecomment-1380908811
But it was like that in your refactor.

Reply:
INstead of using `vscode.tasks.fetchTasks` we can write our own function to read the tasks in tasks.json to fetch the existing tasks, and use the function `provideTasks ` to get the template tasks.
Here is a sample: 

https://github.com/microsoft/vscode-cpptools/blob/a014fe80bdc03b4a9fa3b877866a11d13034f631/Extension/src/LanguageServer/cppBuildTaskProvider.ts#L218

If you need help implementing this let me know and I will modify your branch.

Reply:
I'm not sure in manual parsing of Json is good solution (honestly speaking I am not convinced that ad hock task creation is good solution either).
First of all this will be very fragile in regards to format of Json.
Like I rised in comments - you have added is Template property previously but your code never sets it to true. If you point me what place was intended to have it set then I will add it.

By the way I found one comment from maintainers of VS code which suggests not to use ad hock tasks (https://github.com/microsoft/vscode/issues/59337#issuecomment-519846474) maybe an Original proposal with 'target' retrieval during task execution (instead of having target set by default when task is created). This would also let keep usage of 'shell' task.

By the way I have just discovered that ProblemMatchers stopped working. And this is also related to fact that it is not possible to programicaly define non standard problem matcher in Task constructor.